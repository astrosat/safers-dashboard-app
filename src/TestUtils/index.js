import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { API_PREFIX } from '../api/base';
import { BASE_URL } from '../config';

const middlewares = [thunk];
export const mockStore = configureMockStore(middlewares);

export const baseURL = `${BASE_URL}/${API_PREFIX}`

export const mockedNavigator = jest.fn();
export const mockedHref = jest.fn();
export const mockedSelector = jest.fn();
export const mockedDispatch = jest.fn();