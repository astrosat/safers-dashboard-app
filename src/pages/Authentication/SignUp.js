import React, { useState, useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { Formik } from 'formik';
import { Button, Col, Form, FormGroup, Input, InputGroup, InputGroupText, Label, Row } from 'reactstrap';
import * as Yup from 'yup'
import { signUp as registration, getOrgList, getRoleList } from '../../store/appAction';
import { getGeneralErrors, getError }  from '../../helpers/errorHelper';
import { passwordHelper, pwdRegEx, pwdValidationTxt }  from '../../helpers/passwordHelper';
import { endpoints } from '../../api/endpoints';
import { BASE_URL } from '../../config';
import _ from 'lodash';

const SignUp = () => {
  const [passwordToggle, setPasswordToggle] = useState(false);
  const orgList = useSelector(state => state.common.orgList);
  const roles = useSelector(state => state.common.roleList);
  const error = useSelector(state => state.auth.error);
  const [citizenId, setcitizenId] = useState('');
  const docTNM = BASE_URL +  endpoints.common.termsNconditions;
  const docPP = BASE_URL +  endpoints.common.privacyPolicy;

  if(error){
    window.scrollTo(0, 0);
  }

  // const navigate = useNavigate();
  const dispatch = useDispatch();
  useEffect(() => {
    if(roles.length===0)
      dispatch(getOrgList());
    if(orgList.length===0)
      dispatch(getRoleList());
  }, []);

  useEffect(() => {
    if(roles.length){
      const role = _.find(roles, { name: 'Citizen' });
      setcitizenId(role.id)
    }
  }, [roles]);

  const signUpSchema = Yup.object().shape({
    email: Yup.string()
      .email('Invalid email address')
      .required('The field cannot be empty'),
    password: Yup.string()
      .matches(
        pwdRegEx,
        pwdValidationTxt
      )
      .required('The field cannot be empty'),
    first_name: Yup.string()
      .required('The field cannot be empty'),
    last_name: Yup.string()
      .required('The field cannot be empty'),
    role: Yup.string()
      .required('The field cannot be empty'),
    accepted_terms: Yup.bool()
      .oneOf([true], 'Please accept the terms and conditions')
  }).when((values, schema) => {
    if (values.role !== citizenId) {
      return schema.shape({
        organization: Yup.string().required('The field cannot be empty'),
      });
    }
  });

  return (
    <Formik
      initialValues={{
        email: '',
        first_name: '',
        last_name: '',
        password: '',
        role: '',
        organization: '',
        accepted_terms: false
      }}
      enableReinitialize={true}
      validationSchema={signUpSchema}
      onSubmit={(values, { setSubmitting }) => {
        console.log(values)
        dispatch(registration(values));
        setSubmitting(false);
      }}
      data-test="signUpComponent"
    >
      {({
        values,
        errors,
        touched,
        handleChange,
        handleBlur,
        handleSubmit,
        isSubmitting,
      }) => (
        <div className="jumbotron">
          {getGeneralErrors(error)}
          <div className="container auth-form">
            <Form onSubmit={handleSubmit} noValidate>
              <Row form>
                <Col>
                  <FormGroup className="form-group">
                    <Label for="userEmail">
                      EMAIL ADDRESS:
                    </Label>
                    <Input
                      id="userEmail"
                      className={getError('email', errors, touched)}
                      name="email"
                      placeholder="email address"
                      type="email"
                      onChange={handleChange}
                      onBlur={handleBlur}
                      value={values.email}
                      autoComplete="on"
                      data-testid="sign-up-email"
                    />
                    {getError('email', errors, touched, false)}
                  </FormGroup>
                </Col>
                <Col>
                  <FormGroup className="form-group">
                    <Label for="first_name">
                      FIRST NAME:
                    </Label>
                    <Input
                      id="first_name"
                      className={getError('first_name', errors, touched)}
                      name="first_name"
                      data-testid="sign-up-firstName"
                      placeholder="first name"
                      type="text"
                      onChange={handleChange}
                      onBlur={handleBlur}
                      value={values.first_name}
                      autoComplete="on"
                    />
                    {getError('first_name', errors, touched, false)}
                  </FormGroup>
                </Col>
                <Col >
                  <FormGroup className="form-group">
                    <Label for="last_name">
                      LAST NAME:
                    </Label>
                    <Input
                      id="last_name"
                      className={getError('last_name', errors, touched)}
                      name="last_name"
                      placeholder="last name"
                      type="text"
                      onChange={handleChange}
                      onBlur={handleBlur}
                      value={values.last_name}
                      autoComplete="on"
                      data-testid="sign-up-lastName"
                    />
                    {getError('last_name', errors, touched, false)}
                  </FormGroup>
                </Col>
                <Col>
                  <FormGroup className="form-group">
                    <Label for="userPassword">
                      PASSWORD:
                    </Label>
                    <InputGroup>
                      <Input
                        id="userPassword"
                        className={getError('password', errors, touched)}
                        name="password"
                        placeholder="password"
                        type={passwordToggle ? 'text' : 'password'}
                        onChange={handleChange}
                        onBlur={handleBlur}
                        value={values.password}
                        autoComplete="on"
                        data-testid="sign-up-password"
                      />
                      <InputGroupText>
                        <i data-testid="sign-up-password-toggle" onClick={() => { setPasswordToggle(!passwordToggle) }} className={`fa ${passwordToggle ? 'fa-eye-slash' : 'fa-eye'}`} />
                      </InputGroupText>
                    </InputGroup>
                    {values.password && (<>{passwordHelper(values.password)}</>)}
                    {getError('password', errors, touched, false)}
                  </FormGroup>
                </Col>
                <Col >
                  <FormGroup className="form-group">
                    <Label for="role">
                      SELECT YOUR ROLE:
                    </Label>
                    <Input
                      id="role"
                      className={getError('role', errors, touched)}
                      name="role"
                      placeholder="select role"
                      type="select"
                      onChange={handleChange}
                      onBlur={handleBlur}
                      value={values.role}
                      data-testid="sign-up-role"
                    >
                      <option value={''} >--Select your role--</option>
                      {roles.map((role, index) => { return (<option key={index} value={role.id}>{role.name}</option>) })}
                    </Input>
                    {getError('role', errors, touched, false)}
                  </FormGroup>
                </Col>
                <Col className={values.role == citizenId ? 'd-none': ''}>
                  <FormGroup className="form-group">
                    <Label for="organization">
                      SELECT ORGANISATION:
                    </Label>
                    <Input
                      id="organization"
                      className={getError('organization', errors, touched)}
                      name="organization"
                      placeholder="select organisation"
                      type="select"
                      onChange={handleChange}
                      onBlur={handleBlur}
                      value={values.organization}
                      data-testid="sign-up-org"
                    >
                      <option value={''} >--Select organisation--</option>
                      {orgList.map((org, index) => { return (<option key={index} value={org.id}>{org.name}</option>) })}
                    </Input>
                    {getError('organization', errors, touched, false)}
                  </FormGroup>
                </Col>
                <Col>
                  <FormGroup className="form-group" check>
                    <Input
                      id="accepted_terms"
                      name="accepted_terms"
                      type="checkbox"
                      value={values.accepted_terms}
                      onChange={handleChange}
                      onBlur={handleBlur}
                      data-testid="sign-up-agreeTermsConditions"
                    />
                    <Label
                      check
                      for="accepted_terms"
                    >
                      <p className='mb-0'>
                        <span>I agree to the </span>
                        <a href={docTNM} rel="noreferrer" target="_blank">
                          Terms of User
                        </a>
                        <span> and </span>
                        <a href={docPP} rel="noreferrer" target="_blank">
                          Privacy Policy
                        </a>
                        <span>, to the processing of my personal data, and to receive emails</span>
                      </p>
                    </Label>
                    {getError('accepted_terms', errors, touched, false)}
                  </FormGroup>
                </Col>
              </Row>
              <div className='center-sign-in'>
                <Button
                  className="my-4 sign-in-btn"
                  color="primary"
                  disabled={isSubmitting}
                  data-testid="signUpButton">
                  SIGN UP
                </Button>
              </div>
            </Form>
          </div>
        </div>
      )
      }
    </Formik >
  );
}

export default SignUp;
