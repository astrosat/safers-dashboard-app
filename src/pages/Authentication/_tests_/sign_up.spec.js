import React from 'react';
import { shallow } from 'enzyme';
import { findByTestAttr } from '../../../TestUtils';
import '@testing-library/jest-dom/extend-expect';
import ShallowRenderer from 'react-test-renderer/shallow';
import { act, cleanup, fireEvent, render, screen, waitFor } from '@testing-library/react'
import SignUp from '../SignUp';
import { BrowserRouter } from 'react-router-dom';
// import configureStore from 'redux-mock-store';

const mockedNavigator = jest.fn();
const mockedHref = jest.fn();
const mockedSelector = jest.fn();
const mockedDispatch = jest.fn();

afterEach(cleanup);

jest.mock('react-router-dom', () => ({
  ...(jest.requireActual('react-router-dom')), // technically it passes without this too, but I'm not sure if its there for other tests to use the real thing so I left it in
  useNavigate: () => mockedNavigator,
  useHref: () => mockedHref
}));
jest.mock('react-redux', () => ({
  ...(jest.requireActual('react-redux')), // technically it passes without this too, but I'm not sure if its there for other tests to use the real thing so I left it in
  useSelector: () => mockedSelector,
  useDispatch: () => mockedDispatch
}));
    
describe('Authentication Component', () => {
  let wrapper = null;
  // pay attention to write it at the top level of your file
  describe('Renders', () => {
    beforeEach(() => {
      wrapper = shallow(
        <SignUp/>
      )
    })
  
    it('Should render a authentication container', () => {
      const container = findByTestAttr(wrapper, 'signUpComponent');
      expect(container.length).toBe(1);
    })
    
    it('Should match snapshot', () => {
      const renderer = new ShallowRenderer()
      const result = renderer.render(<SignUp />)
      expect(result).toMatchSnapshot()
    })
      
  });
  
})
describe('Form testing', () => {
  beforeEach(() => {
    render(
      <BrowserRouter>
        <SignUp/>
      </BrowserRouter>)
  })
  test('Should render correctly', () => {
    expect(screen).not.toBeNull();
  })
  test('Fields should be empty initially', () => {
    const emailInput = screen.getByTestId('sign-up-email');
    expect(emailInput).toHaveValue('');
    const firstNameInput = screen.getByTestId('sign-up-firstName');
    expect(firstNameInput).toHaveValue('');
    const lastNameInput = screen.getByTestId('sign-up-lastName');
    expect(lastNameInput).toHaveValue('');
    const roleInput = screen.getByTestId('sign-up-role');
    expect(roleInput).toHaveValue('');
    const orgInput = screen.getByTestId('sign-up-org');
    expect(orgInput).toHaveValue('');
    const passwordInput = screen.getByTestId('sign-up-password');
    expect(passwordInput).toHaveValue('');
    const passwordToggleInput = screen.getByTestId('sign-up-password-toggle');
    expect(passwordToggleInput).toHaveClass('fa-eye')
    const agreeTermsConditions = screen.getByTestId('sign-up-agreeTermsConditions');
    expect(agreeTermsConditions).not.toBeChecked();
  })

  test('Click Toggle password should have change icon', async () => {
    act(() => {
      fireEvent.click(screen.getByTestId('sign-up-password-toggle'))
    });
    let passwordToggleInput = screen.getByTestId('sign-up-password-toggle');
    expect(passwordToggleInput).toHaveClass('fa-eye-slash')
  })
  test('Click Agree to terms and conditions to should set remember me to checked', async () => {
    act(() => {
      fireEvent.click(screen.getByTestId('sign-up-agreeTermsConditions'))
    });
    let agreeTermsConditions = screen.getByTestId('sign-up-agreeTermsConditions');
    await waitFor(() => {
      expect(agreeTermsConditions).toBeChecked();
    }) 
  })
  
})