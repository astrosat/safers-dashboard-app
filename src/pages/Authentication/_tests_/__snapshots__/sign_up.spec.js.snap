// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Authentication Component Renders Should match snapshot 1`] = `
<Formik
  data-test="signUpComponent"
  enableReinitialize={true}
  initialValues={
    Object {
      "accepted_terms": false,
      "email": "",
      "first_name": "",
      "last_name": "",
      "organization": "",
      "password": "",
      "role": "",
    }
  }
  onSubmit={[Function]}
  validationSchema={
    ObjectSchema {
      "_blacklist": ReferenceSet {
        "list": Set {},
        "refs": Map {},
      },
      "_blacklistError": undefined,
      "_excludedEdges": Array [],
      "_nodes": Array [
        "accepted_terms",
        "role",
        "last_name",
        "first_name",
        "password",
        "email",
      ],
      "_sortErrors": [Function],
      "_typeError": [Function],
      "_whitelist": ReferenceSet {
        "list": Set {},
        "refs": Map {},
      },
      "_whitelistError": undefined,
      "conditions": Array [
        Condition {
          "fn": [Function],
          "refs": Array [
            Reference {
              "getter": "",
              "isContext": false,
              "isSibling": false,
              "isValue": true,
              "key": ".",
              "map": undefined,
              "path": "",
            },
          ],
        },
      ],
      "deps": Array [],
      "exclusiveTests": Object {},
      "fields": Object {
        "accepted_terms": BooleanSchema {
          "_blacklist": ReferenceSet {
            "list": Set {},
            "refs": Map {},
          },
          "_blacklistError": undefined,
          "_typeError": [Function],
          "_whitelist": ReferenceSet {
            "list": Set {
              true,
            },
            "refs": Map {},
          },
          "_whitelistError": [Function],
          "conditions": Array [],
          "deps": Array [],
          "exclusiveTests": Object {},
          "spec": Object {
            "abortEarly": true,
            "nullable": false,
            "presence": "optional",
            "recursive": true,
            "strict": false,
            "strip": false,
          },
          "tests": Array [],
          "transforms": Array [
            [Function],
          ],
          "type": "boolean",
        },
        "email": StringSchema {
          "_blacklist": ReferenceSet {
            "list": Set {},
            "refs": Map {},
          },
          "_blacklistError": undefined,
          "_mutate": undefined,
          "_typeError": [Function],
          "_whitelist": ReferenceSet {
            "list": Set {},
            "refs": Map {},
          },
          "_whitelistError": undefined,
          "conditions": Array [],
          "deps": Array [],
          "exclusiveTests": Object {
            "email": false,
            "required": true,
          },
          "spec": Object {
            "abortEarly": true,
            "nullable": false,
            "presence": "required",
            "recursive": true,
            "strict": false,
            "strip": false,
          },
          "tests": Array [
            [Function],
            [Function],
          ],
          "transforms": Array [
            [Function],
          ],
          "type": "string",
        },
        "first_name": StringSchema {
          "_blacklist": ReferenceSet {
            "list": Set {},
            "refs": Map {},
          },
          "_blacklistError": undefined,
          "_mutate": undefined,
          "_typeError": [Function],
          "_whitelist": ReferenceSet {
            "list": Set {},
            "refs": Map {},
          },
          "_whitelistError": undefined,
          "conditions": Array [],
          "deps": Array [],
          "exclusiveTests": Object {
            "required": true,
          },
          "spec": Object {
            "abortEarly": true,
            "nullable": false,
            "presence": "required",
            "recursive": true,
            "strict": false,
            "strip": false,
          },
          "tests": Array [
            [Function],
          ],
          "transforms": Array [
            [Function],
          ],
          "type": "string",
        },
        "last_name": StringSchema {
          "_blacklist": ReferenceSet {
            "list": Set {},
            "refs": Map {},
          },
          "_blacklistError": undefined,
          "_mutate": undefined,
          "_typeError": [Function],
          "_whitelist": ReferenceSet {
            "list": Set {},
            "refs": Map {},
          },
          "_whitelistError": undefined,
          "conditions": Array [],
          "deps": Array [],
          "exclusiveTests": Object {
            "required": true,
          },
          "spec": Object {
            "abortEarly": true,
            "nullable": false,
            "presence": "required",
            "recursive": true,
            "strict": false,
            "strip": false,
          },
          "tests": Array [
            [Function],
          ],
          "transforms": Array [
            [Function],
          ],
          "type": "string",
        },
        "password": StringSchema {
          "_blacklist": ReferenceSet {
            "list": Set {},
            "refs": Map {},
          },
          "_blacklistError": undefined,
          "_mutate": undefined,
          "_typeError": [Function],
          "_whitelist": ReferenceSet {
            "list": Set {},
            "refs": Map {},
          },
          "_whitelistError": undefined,
          "conditions": Array [],
          "deps": Array [],
          "exclusiveTests": Object {
            "matches": false,
            "required": true,
          },
          "spec": Object {
            "abortEarly": true,
            "nullable": false,
            "presence": "required",
            "recursive": true,
            "strict": false,
            "strip": false,
          },
          "tests": Array [
            [Function],
            [Function],
          ],
          "transforms": Array [
            [Function],
          ],
          "type": "string",
        },
        "role": StringSchema {
          "_blacklist": ReferenceSet {
            "list": Set {},
            "refs": Map {},
          },
          "_blacklistError": undefined,
          "_mutate": undefined,
          "_typeError": [Function],
          "_whitelist": ReferenceSet {
            "list": Set {},
            "refs": Map {},
          },
          "_whitelistError": undefined,
          "conditions": Array [],
          "deps": Array [],
          "exclusiveTests": Object {
            "required": true,
          },
          "spec": Object {
            "abortEarly": true,
            "nullable": false,
            "presence": "required",
            "recursive": true,
            "strict": false,
            "strip": false,
          },
          "tests": Array [
            [Function],
          ],
          "transforms": Array [
            [Function],
          ],
          "type": "string",
        },
      },
      "spec": Object {
        "abortEarly": true,
        "nullable": false,
        "presence": "optional",
        "recursive": true,
        "strict": false,
        "strip": false,
      },
      "tests": Array [],
      "transforms": Array [
        [Function],
      ],
      "type": "object",
    }
  }
>
  [Function]
</Formik>
`;
